### DISCUSSION TITLE
Connection successful with Server not able get the Tag values

### DISCUSSION THREAD

**Discussion by Yashvanthb13** [FreeOpcUa/python-opcua#1258] [Q&A]
``` python  
    #!use/bin/env python
    from opcua import Client, ua
    from threading import Thread, Semaphore
    from time import sleep
    from config_read import *
    from datetime import datetime
    import json
    
    
    opcuaTag = {}
    
    sem  = Semaphore()
    
    ServerConfig = GetServerConfig()
    
    ClientConfig = GetClinetConfig()
    
    OpcConfig = GetOpcUaConfig()
    opcuaTag = NodesToRead()
        
    if __name__ == '__main__':
        
        case = 0
        GetAppConfig()
        
          
        URL = OpcConfig['URL']
        
        #initialize the Client object                
        
        while True:
            if  (case == 1):
        
                print('connecting...')   
                try:     
                     
                    client = Client(URL)                          

                    #set the password and User name

                    client.set_user(OpcConfig['username'])          
                    
                    client.set_password(OpcConfig['Password']) 
                    
                    client.application_uri = "urn:freeopcua:client"
                    
                    client.connect()
                                     
                    print("Connection with Client Successful",URL + '\n')
                    
                    print(" Value =", client.get_node("ns=0;i=2267").get_value())
                    
                    case = 2
                
                except:
                    print("Failed to connect to URL", URL)
                    
                            
                    case = 1
                    sleep(2)
                    
            elif (case == 2):
                
                try:  
                    size = opcuaTag['size']
                    
                    Tagvalue = {}
                    
                    for i in range(0 , int(size)):                    
                        Tagvalue["Tag_No_"+ str(i)] = client.get_node(opcuaTag["Variable" + str(i)]).get_value()
                    
                    print(Tagvalue) 
                     
                    print("\n")
                    
                except:
                    
                    print("subcribe Error!")
                    
                    case = 3
                    
            elif (case == 3):
                
                try:
                    service_level = client.get_node("ns=0;i=2267").get_value()
                    
                    if service_level >= 200:
                        
                        case = 3
                        
                    else:
                        
                        case = 4
                        
                    sleep(5)
                    
                except:
                    
                    case = 4
                    
            elif case == 4:
                #disconnect clean = unsubscribe, delete subscription then disconnect
                print("unsubscribing...")
                
                '''
                try:
                    if subscription_handle_list:
                        for handle in subscription_handle_list:
                            subscription.unsubscribe(handle)
                    subscription.delete()
                    print("unsubscribed!")
                except:
                    print("unsubscribing error!")
                    subscription = None
                    subscription_handle_list = []
                    sleep(0)'''
                
                print("disconnecting...")
                
                try:
                    client.disconnect()
                except:
                    print("disconnection error!")
                case = 0
                
            else:
                case = 1
                sleep(2)
            sleep(1) 
```

This the code i'm using to connect and read data from opc ua server, i',m able get data from prosys server simulator,  when i try to connect with siemens 840D server,  connection establishment is successful after that when reading the tag values it is not printing anything  and jumps to print(subcribe Error!).
Server URL and port and other configurations are read from the ini file
 
Able to connect with UaExpert with User name and password  but when i try with above sample code not able to read tag values

**Comment by swamper123** [comment]
(I just made your code example a bit prettier, so that it is better readable. You can do that if you add ```python at the beginning and ``` at the end).
What kind of Exception is thrown?
Does it work if you read a single Node (like in case==3)?

**Reply by Yashvanthb13** [reply]
when using  with username and password, i'm no exception error is thrown, but when i use with certificates  and security policies i'm getting the below errors.

it's showing connection successful

      Received an error: MessageAbort(error:StatusCode(BadSecurityPolicyRejected), reason:None)
      Received an error: MessageAbort(error:StatusCode(BadSecurityPolicyRejected), reason:None)
      Protocol Error
      Traceback (most recent call last):
      File "site-packages\opcua\client\ua_client.py", line 101, in _run
      File "site-packages\opcua\client\ua_client.py", line 121, in _receive
      File "site-packages\opcua\client\ua_client.py", line 129, in _call_callback
      opcua.ua.uaerrors._base.UaError: No future object found for request: 0, callbacks in list are dict_keys([2])


if try to read the single node also it is jumping from that line and printing  subcribe Error!

**Reply by swamper123** [reply]
> This the code i'm using to connect and read data from opc ua server

If username and password is not throwing an exception, then you are not using the code above, do you?

Check the client encryption example:
https://github.com/FreeOpcUa/python-opcua/blob/master/examples/client-with-encryption.py

**Reply by Yashvanthb13** [reply]
i'm using same code for connection with prosys simulator and siemens 840D server aso 

i tired to connect with username and password, with security policy(Basic256Sha256) SecurityMode("SignAndEncrypt") and with certificates also i was getting  the below exception 

        Received an error: MessageAbort(error:StatusCode(BadSecurityPolicyRejected), reason:None)

the above is connecting with prosys server with and without security modes.

**Reply by AndreasHeine** [reply]
you should check if the siemens 840D supports Basic256Sha256 -> "BadSecurityPolicyRejected" this is a pretty precise error!!!

**Reply by Yashvanthb13** [reply]
i tried with different  security policy and mode and with None also but the result is same

**Reply by AndreasHeine** [reply]
UaExpert? does it work? if yes post a screenshot which enpoints are exposed and what configuration is used for connecting!

**Comment by AndreasHeine** [comment]
```python
for i in range(0 , int(size)):                    
                        Tagvalue["Tag_No_"+ str(i)] = client.get_node(opcuaTag["Variable" + str(i)]).get_value()
```

i hope you only read a few nodes with that... many readrequests
https://github.com/FreeOpcUa/python-opcua/blob/0d595e596c9922699f0b36abb79583d5763eeb9b/opcua/client/client.py#L658
is faster and more efficient

using a dict is also slower the using lists or sets

edit:

**POST THE EXCEPTION!**

**Reply by Yashvanthb13** [reply]
20-50 nodes in the loop.

can you provide an example for passing parameters for get_values();

**Reply by AndreasHeine** [reply]
nodes is just a list of node objects??? and the return is a list of values in the same order of the nodeslist!!!

if you look in the sourcecode there is no magic:


```python
    def get_values(self, nodes):
        """
        Read the value of multiple nodes in one roundtrip.
        """
        nodes = [node.nodeid for node in nodes]
        results = self.uaclient.get_attributes(nodes, ua.AttributeIds.Value)
        return [result.Value.Value for result in results]
```

**Reply by AndreasHeine** [reply]
nodes -> list of Node-Class Instances like: [client.get_node(nodeid), client.get_node(nodeid), client.get_node(nodeid), ...] -> readrequest (client.get_values(nodes)) -> readresponse from server -> [val1, val2, val3, ...]

even if new you should know as a baseline what objects, lists, sets, tuples are and how to handle them in order to work with this library?! otherwise you'll have a hard time^^

**Reply by Yashvanthb13** [reply]
Thanks @AndreasHeine  for the example

i'll provide the screenshots
